### DISCUSSION TITLE
How to preserve ISR cache on revalidation error

### DISCUSSION THREAD

**Discussion by KrisStadler** [vercel/next.js#27375] [Help] [answered]
I dont always want to return a 404 page if my relidation function returned invalid page information or the call failed.

In the documentation, it states that if the revalidation function fails the cache will be left alone and the current page will stay online.

I'm struggling to get this to work as the suggested method of throwing an error causes the server to fall over completely

```
export async function getStaticProps() {
    console.log('Get Static Props - making a call');
    let siteDefaults = await getContent('regLoginImage');
    if (siteDefaults?.fields === undefined) {
        console.log('CMS content not found');
         throw Error('CMS content not found');
    }

    console.log('CMS content ready');
    return {
        props: {
            content: siteDefaults 
        }, // will be passed to the page component as props
        revalidate: 10
    };
}
```

**Comment by KrisStadler** [comment, marked as answer]
You ``return: null`` for ``getStaticProps``

**Reply by dipankarmaikap** [reply]
Can you please explain this a bit more or add the full return object.

**Reply by KrisStadler** [reply]
@dipankarmaikap so under normal circumstances getStaticProps returns a object shaped as `{props: yourData}` that will be dropped down to the page via props. 

Given the context I'm using NextJS in (landing page for a marketing system) I can assume that my page will always have content online.

This means that if for some reason the call to fetch the CMS data from `getStaticProps` fails or no content gets returned I want the framework to still serve my previously cached content. Returning `null` for `getStaticProps` results in the framework throwing a warning server-side, but the page will still serve the previously cached content.

On the other hand, if you throw an error your server will crash, and returning an empty object will essentially clear your cached content for that page.

Here is a quick example of how I'm doing this:

```
export async function getStaticProps() {
    try {

	let cmsContent = await getContent();
	
	if (!cmsContent) {
		return null;
	}
	
	return {
		props: {
			content: cmsContent
		}, // will be passed to the page component as props
		revalidate: 10
	};
    } catch(error) {
        console.log(error)
        return null
    }
}
```

Hope this helps :)

**Reply by dipankarmaikap** [reply]
Thanks for the explanation. Lets say for some reason api call failed and I return null will this keep intact  the revalidation .? After 10 sec will this run again.?

Also here is a [comment from](https://github.com/vercel/next.js/discussions/11180#discussioncomment-2399) @Timer recommends not to use try catch in getStaticProps

**Reply by KrisStadler** [reply]
@dipankarmaikap correct, returning null will keep the revalidation intact. your content will also be updated should the API resolve with valid content on the next revalidation run.

 Remember that revalidation is not a timer ie. having a revalidation value of 10 does not mean the page is going to be revalidated every 10 seconds. The revalidation time is simply a time value that indicates how long post-build the framework should wait until it attempts to revalidate a page when loaded.

An example:
Your page has a revalidation value of 10 seconds. You build and deploy your site and within the first 10 seconds someone visits the page, revalidation will not be triggered. If someone lands on the page after 10 seconds only then will revalidation commence. 

You can test this locally by running your project and adding a console log to the getStaticProps method. You should see a console log during the build and then another console log each time you refresh the page twice given that the revalidation time has passed.

Here is a link to some more in detailed docs for ISR: https://vercel.com/docs/next.js/incremental-static-regeneration

**Reply by KrisStadler** [reply]
@dipankarmaikap forgot to mention earlier, thanks for the link to @Timer explanation on why not to user trycatch :)

**Comment by gabrielreisn** [comment]
How is that currently working on app router?? Returning null doesn't avoid it to crash (HTTP 500 internal server error)
